internal/authentication/tokengetter.go

```diff
@@ -0,0 +1,117 @@
+package authentication
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/utils/ptr"
+)
+
+type TokenGetter struct {
+	client            corev1.ServiceAccountsGetter
+	expirationSeconds int64
+	tokens            map[types.NamespacedName]*authenticationv1.TokenRequestStatus
+	tokenLocks        keyLock[types.NamespacedName]
+	mu                sync.RWMutex
+}
+
+type TokenGetterOption func(*TokenGetter)
```

Comment:
When introducing the option pattern to change unexported fields, we should provide some already created option functions to set those values. For example we might add a function called `WithExpirationSeconds` that could be used to set the unexported expirationSeconds field on the TokenGetter

Reply:
Whoops! I commented this before I realized you had already done that - sorry!